File: python/llama-index-server/examples/private_file/custom-workflow.py
from typing import Any, List

from fastapi import FastAPI

from llama_index.core.agent.workflow.workflow_events import AgentStream
from llama_index.core.llms import LLM
from llama_index.core.prompts import PromptTemplate
from llama_index.core.workflow import (
    Context,
    Event,
    StartEvent,
    StopEvent,
    Workflow,
    WorkflowRuntimeError,
    step,
)
from llama_index.llms.openai import OpenAI
from llama_index.server import LlamaIndexServer, UIConfig
from llama_index.server.api.utils.chat_attachments import get_file_attachments
from llama_index.server.models.chat import ChatRequest
from llama_index.server.models.file import ServerFile


class FileHelpEvent(Event):
    """
    The event for helping the user with the an uploaded file.
    """

    file_content: str
    user_request: str


class FileHelpWorkflow(Workflow):
    """
    A simple workflow that helps the user with the an uploaded file.
    Note: The workflow just simply feed all the file content to the LLM so it won't work for large files.
    The purpose is just for demo how a workflow can work with the uploaded file from the user.
    """

    def __init__(
        self,
        llm: LLM,
        file_attachments: List[ServerFile],
        **kwargs: Any,
    ):
        super().__init__(**kwargs)
        self.llm = llm
        self.file_attachments = file_attachments

    @step
    async def read_files(self, ctx: Context, ev: StartEvent) -> FileHelpEvent:
        user_msg = ev.user_msg
        if len(self.file_attachments) == 0:
            raise WorkflowRuntimeError("Please upload one file to start")

        # Read the file content
        last_file = self.file_attachments[-1]
        with open(last_file.path, "r") as f:
            file_content = f.read()

        return FileHelpEvent(
            file_content=file_content,
            user_request=user_msg,
        )

    @step
    async def help_user(self, ctx: Context, ev: FileHelpEvent) -> StopEvent:
        default_prompt = PromptTemplate("""
        You are a writing assistant.
        You are given a file content and a user request.
        Your task is to help the user with the file content.
        
        User request: {user_msg}

        File content:
        {file_content}
        """)
        prompt = default_prompt.format(
            user_msg=ev.user_request,
            file_content=ev.file_content,
        )
        stream = await self.llm.astream_complete(prompt)
        async for chunk in stream:
            ctx.write_event_to_stream(
                AgentStream(
                    response=chunk.text,
                    delta=chunk.delta or "",
                    current_agent_name="agent",
                    tool_calls=[],
                    raw=chunk.raw,
                )
            )

        return StopEvent(
            content=True,
        )


def create_workflow(chat_request: ChatRequest) -> Workflow:
    # Use get_file_attachments to get the file attachments from the chat messages
    file_attachments = get_file_attachments(chat_request.messages)
    return FileHelpWorkflow(
        llm=OpenAI(model="gpt-4.1-mini"),
        file_attachments=file_attachments,
    )


def create_app() -> FastAPI:
    app = LlamaIndexServer(
        workflow_factory=create_workflow,
        suggest_next_questions=False,
        ui_config=UIConfig(
            enable_file_upload=True,
            component_dir="components",
        ),
    )
    return app


app = create_app()


if __name__ == "__main__":
    import uvicorn

    uvicorn.run("custom-workflow:app", host="0.0.0.0", port=8000, reload=True)

File: packages/server/examples/private-file/custom-workflow.ts
import { extractFileAttachments } from "@llamaindex/server";
import { ChatMemoryBuffer, MessageContent, Settings } from "llamaindex";

import {
  agentStreamEvent,
  createStatefulMiddleware,
  createWorkflow,
  startAgentEvent,
  stopAgentEvent,
  workflowEvent,
} from "@llamaindex/workflow";
import { Message } from "ai";
import { promises as fsPromises } from "node:fs";

const fileHelperEvent = workflowEvent<{
  userInput: MessageContent;
  fileContent: string;
}>();

/**
 * This is an simple workflow to demonstrate how to use uploaded files in the workflow.
 */
export function workflowFactory(reqBody: { messages: Message[] }) {
  const llm = Settings.llm;

  // First, extract the uploaded file from the messages
  const attachments = extractFileAttachments(reqBody.messages);

  if (attachments.length === 0) {
    throw new Error("Please upload a file to start");
  }

  // Then, add the uploaded file info to the workflow state
  const { withState, getContext } = createStatefulMiddleware(() => {
    return {
      memory: new ChatMemoryBuffer({ llm }),
      uploadedFile: attachments[attachments.length - 1],
    };
  });
  const workflow = withState(createWorkflow());

  // Handle the start of the workflow: read the file content
  workflow.handle([startAgentEvent], async ({ data }) => {
    const { userInput } = data;
    // Prepare chat history
    const { state } = getContext();
    if (!userInput) {
      throw new Error("Missing user input to start the workflow");
    }
    state.memory.put({ role: "user", content: userInput });

    // Read file content
    const fileContent = await fsPromises.readFile(
      state.uploadedFile.path,
      "utf8",
    );

    return fileHelperEvent.with({
      userInput,
      fileContent,
    });
  });

  // Use LLM to help the user with the file content
  workflow.handle([fileHelperEvent], async ({ data }) => {
    const { sendEvent } = getContext();

    const prompt = `
You are a helpful assistant that can help the user with their file.

Here is the provided file content:
${data.fileContent}

Now, let help the user with this request:
${data.userInput}
`;

    const response = await llm.complete({
      prompt,
      stream: true,
    });

    // Stream the response
    for await (const chunk of response) {
      sendEvent(
        agentStreamEvent.with({
          delta: chunk.text,
          response: chunk.text,
          currentAgentName: "agent",
          raw: chunk.raw,
        }),
      );
    }
    sendEvent(stopAgentEvent.with({ result: "" }));
  });

  return workflow;
}
